### PR TITLE
Hiding digits by default

### PR DESCRIPTION
**Pull request by Da-xy**
The idea of this feature is to have an even more secure mode when entering PIN code.
Feature request linked to this issue : #398 
The code is related to the 4th implementation proposal in the issue.

### CODE REVIEW

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 1]
```diff
@@ -929,8 +942,16 @@ wheel_action_ret_te gui_prompts_render_pin_enter_screen(uint8_t* current_pin, ui
...
+                    sh1122_put_centered_char(&plat_oled_descriptor, PIN_PROMPT_DIGIT_X_OFFS + x_offset_for_digits + PIN_PROMPT_DIGIT_X_SPC*i + PIN_PROMPT_DIGIT_X_ADJ, PIN_PROMPT_UP_ARROW_Y+PIN_PROMPT_ARROW_HEIGHT+PIN_PROMPT_DIGIT_Y_SPACING + anim_step*vert_anim_direction + (PIN_PROMPT_DIGIT_HEIGHT+1)*vert_anim_direction*-1, next_char, TRUE);
+                }
+                else
+                {
+                    sh1122_put_centered_char(&plat_oled_descriptor, PIN_PROMPT_DIGIT_X_OFFS + x_offset_for_digits + PIN_PROMPT_DIGIT_X_SPC*i + PIN_PROMPT_DIGIT_X_ADJ, PIN_PROMPT_UP_ARROW_Y+PIN_PROMPT_ARROW_HEIGHT+PIN_PROMPT_DIGIT_Y_SPACING + extra_y_for_extra_digit + anim_step*vert_anim_direction,  current_char, TRUE);
+                    sh1122_put_centered_char(&plat_oled_descriptor, PIN_PROMPT_DIGIT_X_OFFS + x_offset_for_digits + PIN_PROMPT_DIGIT_X_SPC*i + PIN_PROMPT_DIGIT_X_ADJ, PIN_PROMPT_UP_ARROW_Y+PIN_PROMPT_ARROW_HEIGHT+PIN_PROMPT_DIGIT_Y_SPACING + extra_y_for_extra_digit + anim_step*vert_anim_direction + (PIN_PROMPT_DIGIT_HEIGHT+1)*vert_anim_direction*-1, next_char, TRUE);
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:953]: shouldn't the last arg be FALSE here?

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 2]
```diff
@@ -1123,18 +1145,18 @@ RET_TYPE gui_prompts_get_user_pin(volatile uint16_t* pin_code, uint16_t stringID
     BOOL finished = FALSE;
     
     // Set current pin to 0000 or random
-    if (random_pin_feature_enabled == FALSE)
-    {
-        memset((void*)current_pin, 0, sizeof(current_pin));
-    }
-    else
+    if (show_pin_entry == TRUE && random_pin_feature_enabled == TRUE)
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:1148]: only random when show pin is to TRUE? shouldn't it be the other way around?

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 3]
```diff
@@ -1171,24 +1193,39 @@ RET_TYPE gui_prompts_get_user_pin(volatile uint16_t* pin_code, uint16_t stringID
         /* Click + up/down action */
         if ((detection_result == WHEEL_ACTION_CLICK_UP) || (detection_result == WHEEL_ACTION_CLICK_DOWN))
         {
-            // Do nothing
-            detection_result = WHEEL_ACTION_NONE;
+            // If Click + up
+            if (detection_result == WHEEL_ACTION_CLICK_UP)
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:1197]: you're overriding default behavior here, can you add a if(show_pin_entry)?

**Review by limpkin — CHANGES_REQUESTED** [review]

File: source_code/main_mcu/src/GUI/gui_prompts.c (hunk 3 above)
Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:1197]: You're right! Adding the condition.

File: source_code/main_mcu/src/GUI/gui_prompts.c (hunk 2 above)
Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:1148]: Do you mean changing the order of the condition like this ?
``` C
if (random_pin_feature_enabled == TRUE  && show_pin_entry == TRUE)
```
Or maybe the condition is not correct and I did not understand what your meant ? 😆

File: source_code/main_mcu/src/GUI/gui_prompts.c (hunk 1 above)
Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:953]: I did not get the time to understand completely this function so I do know what the `write_to_buffer` is used for.
I tried setting `FALSE` to those two `sh1122_put_centered_char` calls and the result was that :

The digits are not displaying anymore.

Comment (limpkin) [reply source_code/main_mcu/src/GUI/gui_prompts.c:953]: oh my bad! do not change anything

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 4]
```diff
@@ -722,17 +722,18 @@ static cust_char_t gui_prompts_get_char_to_display(uint8_t const * current_pin,
...
-/*! \fn     gui_prompts_render_pin_enter_screen(uint8_t* current_pin, uint16_t selected_digit, uint16_t stringID, int16_t anim_direction, int16_t vert_anim_direction, int16_t hor_anim_direction, BOOL six_digit_prompt)
+/*! \fn     gui_prompts_render_pin_enter_screen(uint8_t* current_pin, uint16_t selected_digit, uint16_t stringID, int16_t anim_direction, int16_t vert_anim_direction, int16_t hor_anim_direction, BOOL six_digit_prompt, BOOL show_pin)
 *   \brief  Overwrite the digits on the current pin entering screen
 *   \param  current_pin         Array containing the pin
 *   \param  selected_digit      Currently selected digit
 *   \param  stringID            String ID for text query
 *   \param  vert_anim_direction Vertical anim direction (wheel up or down)
 *   \param  hor_anim_direction  Horizontal anim direction (next/previous digit)
 *   \param  six_digit_prompt    TRUE to specify 6 digits prompt
+*   \param  show_pin          TRUE to specify if the PIN digit should be hidden
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:733]: isn't it the contrary? maybe provide a more explicit explanation?

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 5]
```diff
@@ -1171,24 +1193,46 @@ RET_TYPE gui_prompts_get_user_pin(volatile uint16_t* pin_code, uint16_t stringID
...
         {
-            // Do nothing
-            detection_result = WHEEL_ACTION_NONE;
+            if (show_pin_entry) 
+            {
+                // Do nothing
+                detection_result = WHEEL_ACTION_NONE;
+            }
+            else {
+            // If Click + up
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:1202]: { on the line after, please respect indentation

**Review by limpkin — CHANGES_REQUESTED** [review]
nearly there :)

File: source_code/main_mcu/src/GUI/gui_prompts.c (hunk 4 above)
Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:733]: You're right... The variable was named "hide_pin" and I renamed it but forgot to change the commentary.
I changed and add it more details saying that it is FALSE only when the feature "hide pin" is enabled.

File: source_code/main_mcu/src/GUI/gui_prompts.c (hunk 5 above)
Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:1202]: Sure it is fixed now.

File: source_code/main_mcu/src/GUI/gui_prompts.c [hunk 6]
```diff
@@ -1115,6 +1136,7 @@ RET_TYPE gui_prompts_get_user_pin(volatile uint16_t* pin_code, uint16_t stringID
     #endif
     
     BOOL random_pin_feature_enabled = (BOOL)custom_fs_settings_get_device_setting(SETTING_RANDOM_PIN_ID);
+    BOOL show_pin_entry = ((BOOL)custom_fs_settings_get_device_setting(SETTINGS_PARANOID_PIN_ENTRY) == TRUE) ? FALSE : TRUE;
```
Comment (limpkin) [inline source_code/main_mcu/src/GUI/gui_prompts.c:1139]: I should have caught that earlier... but could you do a != FALSE instead, as that value will be programmed by moolticute?
sorry about this additional request :)

Comment (Da-xy) [reply source_code/main_mcu/src/GUI/gui_prompts.c:1139]: No worries!
Indeed it will more relevant to have this condition the other way around :)

**Comment by limpkin** [comment]
thanks a lot :)

**Comment by limpkin** [comment]
@Da-xy : hey there :). I recently noticed that this PR doesn't enable continuous push and scroll... do you have the same issue?

**Comment by limpkin** [comment]
seems that the root cause issue is https://github.com/mooltipass/minible/blame/master/source_code/main_mcu/src/GUI/gui_prompts.c#L814

**Comment by Da-xy** [comment]
Hi @limpkin !
Indeed, I have exactly the same issue !
Unfortunately, I did not get the time to investigate deeper, after the release...
Yes, it seems that the root cause come from this function because removing this line fixes the issue.
However, I am not able to tell if it will bring some sort of side effects if we modify or remove it.
Do you have an idea on a possible fix or do you want me to give it a look ?
Thank you very much for pointing it out !

**Comment by mstephanxanadu** [comment]
I wish you had brought this to our attention earlier, but I should have tested it more extensively.
My idea is to remove this line indeed, and test for potential side effects

**Comment by Da-xy** [comment]
You're right I should have done this before.. sorry about that.
If you want, I can create the related issue with the fix proposal and send a PR with the line removed ?

**Comment by limpkin** [comment]
sorry... had left a comment using another account. 
it's alright, can you make sure removing this line doesn't break anything?

**Comment by Da-xy** [comment]
OK, I will perform some functional tests and I come back to you to inform you if there are any regressions.

**Comment by Da-xy** [comment]
Hi again @limpkin, I checked a lot of different use cases mixing the different options :
- Enabling/Disabling paranoid entry pin
- Enabling/Disabling random entry pin
- Enabling/Disabling display bin on back pressed
- Enabling/Disabling display full pin on entry

I also checked with the changing pin display.
I did not notice any issues with the fix. So I think, that this instruction can be deleted.

However, I felt that mixing these options are not very good and could lead to a very weird user experience.
I think that I will create an issue on Moolticute repository in order to submit a feature request. It will prevent new users to select all options which might in my opinion will make no sense especially : paranoid entry pin + display full pin on entry.

**Pull request merged**